state/pruner.go

```diff
@@ -0,0 +1,267 @@
+		} else {
+			return err
+		}
+	}
+	currentAppRetainHeight, err := p.stateStore.GetApplicationRetainHeight()
+	var noAppRetainHeight bool
+	if err != nil {
+		if err == ErrKeyNotFound {
+			noAppRetainHeight = true
+		} else {
+			return err
+		}
+	}
+	if currentCompanionRetainHeight > height || (!noAppRetainHeight && currentAppRetainHeight > height) {
+		return errors.New("cannot set a height lower than previously requested - blocks might have already been pruned")
+	}
+	err = p.stateStore.SaveCompanionBlockRetainHeight(height)
+	return err
+}
+
+// SetABCIResRetainHeight sets the retain height for ABCI responses.
+//
+// If the application has set the DiscardABCIResponses flag to true, nothing
+// will be pruned.
+func (p *Pruner) SetABCIResRetainHeight(height int64) error {
+	if height <= 0 || height > p.bs.Height() {
+		return ErrInvalidHeightValue
+	}
+	currentRetainHeight, err := p.stateStore.GetABCIResRetainHeight()
+	if err != nil {
+		if err == ErrKeyNotFound {
+			err = p.stateStore.SaveABCIResRetainHeight(height)
+			return err
+		}
+		return err
+	}
+	if currentRetainHeight > height {
+		return errors.New("cannot set a height lower than previously requested - blocks might have already been pruned")
+	}
+	err = p.stateStore.SaveABCIResRetainHeight(height)
+	return err
+}
+
+func (p *Pruner) pruningRoutine() {
+	p.logger.Info("Pruner started", "interval", p.interval.String())
+	lastHeightPruned := int64(0)
+	lastABCIResPrunedHeight := int64(0)
+	for {
+		select {
+		case <-p.Quit():
+			return
+		default:
+			lastHeightPruned = p.pruneBlocksToRetainHeight(lastHeightPruned)
+			lastABCIResPrunedHeight = p.pruneABCIResToRetainHeight(lastABCIResPrunedHeight)
+			time.Sleep(p.interval)
+		}
+	}
+}
+
+func (p *Pruner) pruneBlocksToRetainHeight(lastHeightPruned int64) int64 {
+	retainHeight := p.findMinRetainHeight()
+	if retainHeight != lastHeightPruned {
+		pruned, evRetainHeight, err := p.pruneBlocks(retainHeight)
+		if err != nil {
+			p.logger.Error("Failed to prune blocks", "err", err, "blockRetainHeight", retainHeight)
+		} else if pruned > 0 {
+			p.logger.Info("Pruned blocks", "count", pruned, "evidenceRetainHeight", evRetainHeight)
+		}
+		lastHeightPruned = retainHeight
+	}
+	return lastHeightPruned
+}
+
+func (p *Pruner) pruneABCIResToRetainHeight(lastABCIResPrunedHeight int64) int64 {
+	abciResRetainHeight, err := p.stateStore.GetABCIResRetainHeight()
+	if err != nil {
+		// ABCI response retain height has not yet been set - do not log any
+		// errors at this time.
+		if errors.Is(err, ErrKeyNotFound) {
+			return 0
+		}
+		p.logger.Error("Failed to get ABCI result retain height", "err", err)
+	} else if lastABCIResPrunedHeight != abciResRetainHeight {
+		prunedHeight, err := p.stateStore.PruneABCIResponses(abciResRetainHeight)
+		if err != nil {
+			p.logger.Error("Failed to prune ABCI responses", "err", err, "abciResRetainHeight", abciResRetainHeight)
+		}
+		p.logger.Info("Pruned ABCI responses", "height", prunedHeight)
```

Comment:
Here we miss setting `lastABCIPrunedHeight` to `abciResRetainHeight` as we are doing in the method above for `lastHeightPruned`